### DISCUSSION TITLE
Argument of type 'StateCreator<..., [], [["zustand/devtools", never], ["zustand/persist", ...]]>' is not assignable to parameter of type 'StateCreator<..., [], [never, unknown][]>'.

### DISCUSSION THREAD

**Discussion by keduong33** [pmndrs/zustand#2168] [Q&A] [answered]
Hi everyone,

I was implementing Zustand to my new project but I could not get it to work (including the [example](https://docs.pmnd.rs/zustand/guides/typescript#using-middlewares) given by the documents). It gives me this error (I took out devtools to make it a bit easier to read):

```text
Argument of type 'StateCreator<BearState, [], [["zustand/persist", BearState]]>' is not assignable to parameter of type 'StateCreator<BearState, [], [never, unknown][]>'.
  Type 'StateCreator<BearState, [], [["zustand/persist", BearState]]>' is not assignable to type '{ $$storeMutators?: [never, unknown][] | undefined; }'.
    Types of property '$$storeMutators' are incompatible.
      Type '[["zustand/persist", BearState]] | undefined' is not assignable to type '[never, unknown][] | undefined'.
        Type '[["zustand/persist", BearState]]' is not assignable to type '[never, unknown][]'.
          Type '["zustand/persist", BearState]' is not assignable to type '[never, unknown]'.
```
           
The code I tried to implement:

```typescript
const useBearStore2 = create<BearState>()(
    persist(
      (set) => ({
        bears: 0,
        increase: (by) => set((state) => ({ bears: state.bears + by })),
      }),
      { name: "bearStore" }
  )
);
```
---

I tried another method without workaround [docs](https://docs.pmnd.rs/zustand/guides/typescript#create-without-curried-workaround)
```typescript
const useBearStore = create<BearState, [["zustand/persist", BearState]]>(
  persist(
    (set) => ({
      bears: 0,
      increase: (by) => set((state) => ({ bears: state.bears + by })),
    }),
    { name: "bearStore" }
  )
);
```

It gave me this error:
```
Type '[["zustand/persist", BearState]]' does not satisfy the constraint '[never, unknown][]'.
  Type '["zustand/persist", BearState]' is not assignable to type '[never, unknown]'.
    Type at position 0 in source is not compatible with type at position 0 in target.
      Type 'string' is not assignable to type 'never'.ts(2344)
```
---
I could not replicate the issue on TS playground: [link](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgYygUwIYzQGjjATzDTgGUYs0BhdLaOAXzgDMoIQ4AiALwFcBnCgDsAJpwDcAKFCRYiOCLQA3GBAgAbfnmJR+wQYxZsOPAcJEB6EMBEj1aAO4Z0EydKHYozDMhIAhTChySkRJODgAI0D+AC44IV4QKKgpcOAhVEx+NDiACgiCOISktCgASjgAXgA+OCUIGykGN2QIIQMBNADncmgSSpRabAAeMLhuoIpsHDGAbVnTQQxRC0UVNU1OPCFlUoBdPAW+JZWdPUEt8cDg7D29yWrcsbXVDX4n8PCz-RgPz-DctkYBUanBcggxv9wlFnLE4AAGGZQz7pTIYbJ5Aog2pA3KAqZobFgpAw3RxJbYAB0pP4cAA1JECIwymUkVCGKzIZ8kEIMCAclxSb0XIwuWUxuKylJJK12vBOhNhWgAExVQaYEaKgmPMp-F4bd5c76CP7-QFoYFVWrgrn-GlxRG2lEZWgYsFYq1wXH4yhE8GRaLkgnU6L0xnMznIjls-48vkCzhC1Qi5r-cXhSXS2UdbJavoAZjiN2oQ2go3CWsobPmi3Mq2Ur02212UAOcCOZmWlmNMEuldu91qAzRmuu2tyuueDYNprgPdnAKBfohyOhgYRMahqNdAvyBCJ3ophM9-vtXuDNLDBQjm-C0adcf5cUTY76nFFaYlkilkiAA)

---

This is my repo: [link](https://github.com/keduong33/zustand-error.git)

### Step to reproduce:
`git clone https://github.com/keduong33/zustand-error.git`
`npm install` 
Go to Bear.store.ts, the stores should be all underlined red.

**Comment by keduong33** [comment]
I digged a bit more but I found another similar problem: https://github.com/pmndrs/zustand/discussions/2164 --> Potential duplicates

**Comment by nfnfgo** [comment]
I think i has the similar problem 

and I tried to completely copy the code from the offical doc


```tsx
import { create } from 'zustand'
import { immer } from 'zustand/middleware/immer'

type State = {
    count: number
}

type Actions = {
    increment: (qty: number) => void
    decrement: (qty: number) => void
}

export const useCountStore = create<State & Actions>()(
    immer((set) => ({
        count: 0,
        increment: (qty: number) =>
            set((state) => {
                state.count += qty
            }),
        decrement: (qty: number) =>
            set((state) => {
                state.count -= qty
            }),
    }))
)
```

But typescript complains there is an error

![image](https://github.com/pmndrs/zustand/assets/61616918/241ba733-c904-4138-aedc-1aa4f115b6b3)
![image](https://github.com/pmndrs/zustand/assets/61616918/a61ca6d4-df03-404b-8b3d-379fb8ecad1c)

**Reply by nfnfgo** [reply]
I also use the same pattern (zustand + immer middleware) in my previous project, and i found the older project doesn't have this issue, so I guess maybe it's a issue in newer version of zustand or sth else

**Comment by pablohaller** [comment]
Hey! I think it might be a problem with VSCode, or with the configuration from tsconfig, or eslint, a NextJS config, I don't know.
If you do a // @ts-nocheck it will probably work right away, so keep it in mind, don't let that problem distract you!
I'm trying to find a solution for that too!

**Comment by dbritto-dev** [comment, marked as answer]
Hey peeps, there's a fix for that https://github.com/pmndrs/zustand/pull/2170. BTW, would you mind trying the local instructions (https://ci.codesandbox.io/status/pmndrs/zustand/pr/2170/builds/434064)?

**Reply by keduong33** [reply]
Yup following the local instructions fixed my problem. In case anyone missed it: `npm i https://pkg.csb.dev/pmndrs/zustand/commit/992cedd4/zustand`. I think #2170 has more info on what is going on.

**Reply by code1379** [reply]
Thank you very much! It worked after I ran `npm i https://pkg.csb.dev/pmndrs/zustand/commit/992cedd4/zustand`.
installed zustand version 4.4.5, then upgraded it. 
Now, there are no more TypeScript errors. I'm not sure why it worked, but it did!

**Comment by mrbeardad** [comment]
Still have this problem
![image](https://github.com/pmndrs/zustand/assets/54128430/508fc775-1557-48a9-be9d-7e56b1181ee2)

**Reply by dbritto-dev** [reply]
@mrbeardad  would you mind sharing us a ts playground?

**Reply by keduong33** [reply]
Hey did you mean to reply to the other person cause we resolved this a while ago? @dbritto-dev

**Reply by dbritto-dev** [reply]
@keduong33 my bad, was the autocomplete hehe

**Reply by mrbeardad** [reply]
@dbritto-dev Maybe it's because my node version is too low, it works good on my home pc, ignore the issue.

**Reply by KirDontsov** [reply]
Same here 
<img width="1130" alt="Снимок экрана 2025-03-14 в 00 41 54" src="https://github.com/user-attachments/assets/b82dfa60-8f66-438c-837a-6b8122af8f64" />

**Reply by dbritto-dev** [reply]
@KirDontsov tsplaground?

**Reply by sandrig** [reply]
Similar problem
"immer": "10.1.1",
"zustand": "5.0.4"
"typescript": "5.8.3"

NodeJS v20.18.3

<img width="1111" alt="Screenshot 2025-05-17 at 18 51 33" src="https://github.com/user-attachments/assets/f9bc07cc-4ab1-4ebb-9f09-7bf6a62780fa" />

### Solution

```ts
// Store (add export type AlertsStoreState)
export type AlertsStoreState = AlertsSlice & AlertSlice

const useAlertsStore = create<AlertsStoreState>()(
persist(
    immer((...a) => ({
      ...createAlertsSlice(...a),
      ...createAlertSlice(...a),
    })),
   {
      name: 'alerts-storage',
      partialize: (state) => ({
        listNameFilter: state.listNameFilter,
        listPageSize: state.listPageSize,
        listSorting: state.listSorting,
      }),
      version: 1,
    },
  ),
)

// Slice (add all state AlertsStoreState)
export const createAlertSlice: StateCreator<AlertsStoreState, [['zustand/immer', never]], [], AlertSlice> = (
  set,
  get,
) => ({
  // ....
})

```

**Reply by dbritto-dev** [reply]
@KirDontsov @sandrig you should use `create()(...)` or `createStore()(...)` always unless you use js without middlewares then you should use `create(...)` or `createStore(...)` please keep in mind that Zustand docs are js first unless you see types.

The new docs apis, hooks and middlewares are more ts first

**Reply by RajuGangitla** [reply]
same issue for me also 
did anyone got solution for this one ?
![image](https://github.com/user-attachments/assets/26628e79-25a7-4f00-974c-5799fafd133a)


![image](https://github.com/user-attachments/assets/d40e7eda-e223-4ea2-84fd-86aba0c5234c)

**Reply by dbritto-dev** [reply]
@RajuGangitla ts playground or stackblitz minimal repro?

**Comment by webmonch** [comment]
I my case the issue was gone after I restarted VS Code

**Reply by riicardojs** [reply]
Tks

**Comment by gentoo90** [comment]
Just stumbled upon this. In my case it was caused by `"resolvePackageJsonImports": false` in `tsconfig.json`.
